IR-check info

sjrd opened the pull request.

sjrd commented:
CI only for now.

scala-jenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/1304/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/1304/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/1644/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/1644/</a>
Test PASSed.

scala-jenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/1303/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/1303/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/1643/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/1643/</a>
Test FAILed.

scala-jenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/1305/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/1305/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/1645/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/1645/</a>
Test FAILed.

scala-jenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/1306/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/1306/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/1646/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/1646/</a>
Test FAILed.

scala-jenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/1307/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/1307/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/1647/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/1647/</a>
Test FAILed.

scala-jenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/1309/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/1309/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/1649/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/1649/</a>
Test PASSed.

scala-jenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/1308/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/1308/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/1648/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/1648/</a>
Test FAILed.

scala-jenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/1311/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/1311/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/1652/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/1652/</a>
Test PASSed.

scala-jenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/1312/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/1312/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/1653/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/1653/</a>
Test PASSed.

sjrd commented:
It should be complete now.
Review by @gzm0 if you get some time this week.

sjrd commented:
partest --fastOpt succeeded locally, except for the 3 failures already in master filed as #1434.

scala-jenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/1310/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/1310/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/1651/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/1651/</a>
Test FAILed.

gzm0 commented:
LGTM unitl 71fc4ba4aaa32398c87685ef724df5811b2169b8

gzm0 commented on ir/src/main/scala/org/scalajs/core/ir/Infos.scala:
```diff
@@ -64,4 +70,292 @@ object Infos {
+           * always kept anyway, because it is in scalajsenv.js.
+           * Other methods delegate to Object, which we can model with
+           * a static call to Object.method.
+           */
```
This is conceptually wrong. What is in the `scalajsenv.js` and what not should be internal to the linker and not influence binary compatibility.

sjrd replied:
The problem is that we don't have any reification of the `Array` class family in the info. We have them in types as `ArrayType`s, but not in the info.
Before this commit the situation was even wrong: if you created a small hello-world problem that only calls `notify()` on an `Array` (but on nothing else), the `Analyzer` would drop it.

It may be conceptually wrong, but at least it will never change. Whether we implement these things in `scalajsenv.js` or not, it will always be the case that all methods but `clone__O` dispatch to `Object`. And `Array.clone` is so weird that it can never be written in user-space.

If you have a better idea, I'm open. That's the best I could come up with.

gzm0 commented on ir/src/main/scala/org/scalajs/core/ir/Infos.scala:
```diff
@@ -64,4 +70,292 @@ object Infos {
+          addMethodCalledStatically(ObjectClass, method)
+
+        case _ =>
+          // Nothing to do
```
In what cases do (should) we get here?

sjrd replied:
`Null` and `Nothing`. There's also `RecordType` but that shouldn't happen.

gzm0 commented on ir/src/main/scala/org/scalajs/core/ir/Infos.scala:
```diff
@@ -64,4 +70,292 @@ object Infos {
+    }
+
+    def addAccessedClassData(cls: String): this.type = {
+      if (!AlwaysPresentClassData.contains(cls))
```
Again, this should be private to the linker.

sjrd replied:
Well it's just part of the spec of the info that this precise set of classes is never mentioned. The set is not arbitrary: it is tightly coupled with the semantics of Scala.js. There are other class data never eliminated by the linker that are not mentioned here, because those are artifacts of the linker. But `Object`, primitive classes and hijacked classes must necessarily always be present due to the semantics.

An alternative, if you still don't like it, is to exclude only `PrimitiveClasses` here (the `Analyzer` would not be happy if it received an `I` here).

sjrd replied:
Or make it a precondition that one cannot call this method with one of the `PrimitiveClasses`.

sjrd replied:
This was also prompted by the fact that an `Is/AsInstanceOf(x, ClassType("jl_Integer"))` does not actually need the data of `jl_Integer`. This is the case for all hijacked classes. But I could also special-case the hijacked boxed classes in `Is/AsInstanceOf` and short-circuit them there. Maybe that's better. And then improve the compiler a tiny little bit to avoid generating `AsInstanceOf(x, ClassType("O"))`, which are no-ops, and are the main source of accessed-class-data "O".

gzm0 commented on ir/src/main/scala/org/scalajs/core/ir/Infos.scala:
```diff
@@ -64,4 +70,292 @@ object Infos {
+   *  record them.
+   */
+  private val AlwaysPresentClassData =
+    PrimitiveClasses ++ HijackedClasses + ObjectClass
```
Knowledge internal to linker.

gzm0 commented:
That's all until e0c30db

sjrd commented:
Updated. I added the commit "Avoid to emit `.is/asInstanceOf[Object]`", and in the commit "Add builders for Infos" I changed only `Infos.scala`.

scala-jenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/1318/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/1318/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/1660/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/1660/</a>
Test PASSed.

scala-jenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/1320/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/1320/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/1663/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/1663/</a>
Test PASSed.

gzm0 commented on tools/shared/src/main/scala/org/scalajs/core/tools/optimizer/InfoChecker.scala:
```diff
@@ -0,0 +1,147 @@
+        info.isExported != expectedInfo.isExported ||
+        info.kind != expectedInfo.kind ||
+        info.superClass != expectedInfo.superClass ||
+        info.interfaces != expectedInfo.interfaces) {
```
Does the order actually matter here?

sjrd replied:
Hum no it doesn't.

gzm0 commented:
The `InfoChecker` should really not be invoked from inside the linker. The `ScalaJSOptimizer` should handle the pipeline dispatch. I think the `InfoChecker` should be made part of the IR checker. This is especially important since phases might bring the infos and the IR out of sync. So it is crucial to be able to check between (future) phases.

I don't think its a big deal to not check the infos of non-reachable classes, since their IR might be useless anyway.

gzm0 commented:
That's all

sjrd commented:
> I don't think its a big deal to not check the infos of non-reachable classes, since their IR might be useless anyway.

On the contrary, it is crucial, since infos decide what is reachable and what is not. If the info are wrong, you could miss reachable classes or methods that you need to check the info for!

sjrd commented:
Updated.

* Rebased on top of master (with 2.11.5)
* Use the info builders also for the manual `java.lang.Object` and `java.lang.String`, as well as for the partest launcher object.
* Fixed the problem with checking the order of interfaces.

sjrd commented:
And partest --fastOpt succeeded locally.

sjrd commented:
OK finally I found a solution to the `Array` problem: in the info, the `ArrayType`s are represented by a pseudo `s_Array` class (encoding of `scala.Array`). This pseudo class is hijacked by the `Analyzer` so that it applies the hack with `clone__O`/others. This way there is no hack in the .sjsir files, and only the `Analyzer` needs to know about this problem.

gzm0 commented:
Ok. Cool. Let's leave the InfoChecker as it is for now. We'll have to allow it between phases once we allow plugins. (We are used to accumulate debt in this thing anyway :P).

gzm0 commented:
LGTM

scala-jenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/1335/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/1335/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/1682/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/1682/</a>
Test PASSed.

sjrd commented:
Cool :)
